tests/e2e/utils/data_fetcher/traces.py

```diff
@@ -0,0 +1,266 @@
+        self.metadata: List[Dict[str, Dict[str, Any]]] = []
+        self.minimum_traces = minimum_traces
+
+        paginator = self.xray_client.get_paginator("get_trace_summaries")
+        pages = paginator.paginate(
+            StartTime=self.start_date,
+            EndTime=self.end_date,
+            TimeRangeType="Event",
+            Sampling=False,
+            FilterExpression=self.filter_expression,
+        )
+
+        self.trace_ids = self._get_trace_ids(pages)
+        self.trace_documents = self._get_trace_documents()
+        self.subsegments = self._get_subsegments()
+
+    def get_annotation(self, key: str, value: Optional[any] = None) -> List:
+        return [
+            annotation
+            for annotation in self.annotations
+            if (value is not None and annotation.get(key) == value) or (value is None and key in annotation)
+        ]
+
+    def get_metadata(self, key: str, namespace: str = "") -> List[Dict[str, Any]]:
+        seen = []
+        for meta in self.metadata:
+            metadata = meta.get(namespace, {})
+            if key in metadata:
+                seen.append(metadata)
+        return seen
+
+    def get_subsegment(self, name: str) -> List:
+        return [seg for seg in self.subsegments if seg.name == name]
+
+    def _find_nested_subsegments(self, subsegment: TraceSubsegment, seen: List):
+        """Recursively add any subsegment that we might be interested.
+
+        It excludes any subsegments contained in exclude_segment_name.
+        Since these are nested, subsegment name might be '## lambda_handler'.
+
+        Parameters
+        ----------
+        subsegment : TraceSubsegment
+            subsegment to traverse
+        seen : List
+            list of subsegments to be updated
+        """
+        if subsegment.subsegments is not None:
+            for seg in subsegment.subsegments:
+                if seg.name not in self.exclude_segment_name:
+                    seen.append(seg)
+                if seg.annotations:
+                    self.annotations.append(seg.annotations)
+                if seg.metadata:
+                    self.metadata.append(seg.metadata)
+
+        return seen
+
+    def _get_subsegments(self) -> List[TraceSubsegment]:
+        """Find subsegments and potentially any nested subsegments
+
+        It excludes any subsegments contained in exclude_segment_name.
+        Since these are top-level, subsegment name might be 'Overhead/Invocation, etc.'.
+
+        Returns
+        -------
+        List[TraceSubsegment]
+            List of subsegments
+        """
+        seen = []
+        for document in self.trace_documents.values():
+            if document.subsegments:
+                for subsegment in document.subsegments:
+                    if subsegment.name not in self.exclude_segment_name:
+                        seen.append(subsegment)
+                    self._find_nested_subsegments(subsegment, seen)
```

Comment:
If I read this correctly, this one will find only directly nested subsegments. Should we also account for a recursive nested subsegments ( not only in first branch, but also in second/third)?

Reply:
true! Need to think of a cleaner way

Reply:
finally found it! Here's the solution using generator. [`15b3285` (#1457)](https://github.com/awslabs/aws-lambda-powertools-python/pull/1457/commits/15b3285318204e28e7f73c5992df3cd3246b524c)

Updating tests to make all of them use nested subsegments to cover them too - making way more robust too.

**PS**: That was challenging!!

![image](https://user-images.githubusercontent.com/3340292/185365711-aec094cd-10fb-40ad-b65c-91c77a811580.png)

![image](https://user-images.githubusercontent.com/3340292/185365782-f44785a6-2a84-44fb-ba91-851cd80c0394.png)